relay-redis/src/real.rs

```diff
@@ -1,146 +1,237 @@
+            }
+        };
+
+        Ok(connection)
     }
 
-    /// Return [`RedisClientStats`] for [`AsyncRedisClient`].
+    /// Returns statistics about the current state of the connection pool.
     ///
-    /// It will always return 0 for `idle_connections` and 1 for `connections` since we
-    /// are re-using the same connection.
+    /// Provides information about the number of active and idle connections in the pool,
+    /// which can be useful for monitoring and debugging purposes.
     pub fn stats(&self) -> RedisClientStats {
+        let status = match self {
+            Self::Cluster(pool) => pool.status(),
+            Self::Single(pool) => pool.status(),
+        };
+
         RedisClientStats {
-            idle_connections: 0,
-            connections: 1,
+            idle_connections: status.available as u32,
+            connections: status.size as u32,
+            max_connections: status.max_size as u32,
+            waiting_for_connection: status.waiting as u32,
         }
     }
-}
 
-/// A wrapper type for async redis connections.
-#[derive(Clone)]
-pub enum AsyncRedisConnection {
-    /// Variant for [`ClusterConnection`].
-    Cluster(ClusterConnection),
-    /// Variant for [`ConnectionManager`] using [`redis::aio::MultiplexedConnection`].
-    Single(ConnectionManager),
-}
+    /// Runs the `predicate` on the pool blocking it.
+    ///
+    /// If the `predicate` returns `false` the object will be removed from pool.
+    pub fn retain(&self, mut predicate: impl FnMut(Metrics) -> bool) {
+        match self {
+            Self::Cluster(pool) => {
+                pool.retain(|_, metrics| predicate(metrics));
+            }
+            Self::Single(pool) => {
+                pool.retain(|_, metrics| predicate(metrics));
+            }
+        }
+    }
 
-impl AsyncRedisConnection {
-    /// Creates an [`AsyncRedisConnection`] in cluster mode.
-    pub async fn cluster<'a>(
-        servers: impl IntoIterator<Item = &'a str>,
+    /// Builds a [`Pool`] given a type implementing [`Manager`] and [`RedisConfigOptions`].
+    fn build_pool<M: Manager + 'static, W: From<Object<M>> + 'static>(
+        manager: M,
         opts: &RedisConfigOptions,
-    ) -> Result<Self, RedisError> {
-        // connection timeout is set in `base_pool_builder` on the pool level
-        let client = ClusterClientBuilder::new(servers)
-            .response_timeout(Duration::from_secs(opts.read_timeout))
-            .connection_timeout(Duration::from_secs(opts.connection_timeout))
-            .build()
-            .map_err(RedisError::Redis)?;
-        let connection = client
-            .get_async_connection()
-            .await
-            .map_err(RedisError::Redis)?;
-        Ok(Self::Cluster(connection))
+    ) -> Result<Pool<M, W>, BuildError> {
+        let result = Pool::builder(manager)
+            .max_size(opts.max_connections as usize)
+            .create_timeout(opts.create_timeout.map(Duration::from_secs))
```

Comment:
What happened to the wait timeout here? I thought we discussed making it an optional config value but default it to `None`.

Reply:
My bad, I thought we agreed on not having it because we wanted to avoid people configuring it because it leads to data loss.